godot_fmod.cpp

```diff
@@ -41,39 +42,27 @@ void Fmod::init(int numOfChannels, int studioFlags, int flags) {
 }
 
 void Fmod::update() {
-	// clean up one shots
-	for (int i = 0; i < oneShotInstances.size(); i++) {
-		auto instance = oneShotInstances.get(i);
-		FMOD_STUDIO_PLAYBACK_STATE s;
-		checkErrors(instance->getPlaybackState(&s));
-		if (s == FMOD_STUDIO_PLAYBACK_STOPPED) {
-			checkErrors(instance->release());
-			oneShotInstances.remove(i);
-			i--;
-		}
-	}
-
-	// update and clean up attached one shots
-	for (int i = 0; i < attachedOneShots.size(); i++) {
-		auto aShot = attachedOneShots.get(i);
-		if (isNull(aShot.gameObj)) {
-			FMOD_STUDIO_STOP_MODE m = FMOD_STUDIO_STOP_IMMEDIATE;
-			checkErrors(aShot.instance->stop(m));
-			checkErrors(aShot.instance->release());
-			attachedOneShots.remove(i);
-			i--;
-			continue;
-		}
-		FMOD_STUDIO_PLAYBACK_STATE s;
-		checkErrors(aShot.instance->getPlaybackState(&s));
-		if (s == FMOD_STUDIO_PLAYBACK_STOPPED) {
-			checkErrors(aShot.instance->release());
-			attachedOneShots.remove(i);
-			i--;
-			continue;
-		}
-		updateInstance3DAttributes(aShot.instance, aShot.gameObj);
-	}
+	for (auto e = events.front(); e; e = e->next()) {
+		FMOD::Studio::EventInstance *eventInstance = e->get();
+        EventInfo *eventInfo = getEventInfo(eventInstance);
+        if (eventInfo->isOneShot) {
```

Comment:
BTW, I know this is was the original logic of the previous update() method. But you do not have to check for the playback state in order to call `release()` on a one-shot event. 

You can issue the release call immediately after a startEvent call. This schedules it for cleanup and FMOD will automatically collect it after the event finishes playing.

Reply:
So you basically mean that we only need to keep track on attached events ?

Reply:
We still need to hold a reference to oneShot events in order to use function as MuteAllEvents Or PauseAllEvents. This reference has to be removed when the Oneshot is terminated so we still need to check its state. We can't do that if the event instance has already been removed from the memory.

Reply:
Well I suppose it is a problem we created ourselves since we iterate through all idnividual event instead of using directly the masterbus

Reply:
Ah yes forgot about that.

Reply:
Come to think of it, we are adding so much complexity into the class because of the mute and pause feature. Perhaps there is a simpler way to handle all of it?

Reply:
We didn't add the EventInfo Struct only for the mute/pause feature, but to have a simple way to attach data to an EventInstance. We believe that it will pay in the future when more functionnalities will be added.
The thing is that EventInstance doesn't have a mute option at all, that why we emulate that by setting the volume to 0 and caching the old volume (instead of resetting to 1 when unmuted which can be incorrect).
The real question is: Do we wish to pause/mute at the EventInstance level or Bus Level when choosing "All"?

Reply:
Not quite sure. It was @piiertho's idea. Because this allows you to mute all events and unmute one by one. It's a valid but quite rare of a scenario. Muting at the master bus hierarchy certainly makes things a lot simpler. 

We could potentially make a judgement call here and extract this out of the integration all together. Users can mute at the instance level and maintain the old volume levels in script.

Thoughts?

Reply:
I am not fond of the idea. A "simple" feature like preserving the initial volume should not be user responsability.
We are talking implementation detail here. Yes, muting/pausing the bus is much simpler right now. But will it not introduce side effects, if we later wish to add bus control functionnalities to the integration?
I am not against the idea, I am just cautious.

Reply:
I prefer the cpp way, because it seems to be an integration feature. To me, we only need GDScript bindings to keep it simple for designers. Another problem with that idea is that it extracts a feature from the build, extracting it out of binary to a script.
I really wish we can accord on features that should be implemented in each version, in the future. So that we would be able to unify parts of our Docs, provide same functionality to users for each release, etc.

Reply:
> A "simple" feature like preserving the initial volume should not be user responsability.

IMO, I would consider this one to be a "nice to have" feature. The FMOD API by default doesn't let you mute individual instances. You would have to do it with parameters in Studio (this would be the ideal workflow for a designer either way).

> But won't it introduce side effects if we later wish to add bus control functionnalities to the integration?

It's highly unlikely as we are offloading all the work to the FMOD Core without maintaining it ourselves. 


All in all, I'm not totally against this feature. Like I mentioned, it's a "nice to have". The only thing I'm concerned about is for a "nice to have" feature it's taking up a pretty large chunk of code to implement.

Reply:
I fail to see where there is a lot of changes involving the mute fonctionnality. 
We juste added isMuted and oldVolume to the EventInfo struct then we used those two variables in the Mute/UnMute functions.
All others changes are part of are working we made to simplify the code. (No more single map to handle OneShot and Attached event therefore a simpler update function).

Reply:
On the other side, I agree that this feature is outside of the FMOD API.
Our initial reasonning with the "old volume caching" was your first implementation of UnMuteAll() where you was setting the volume to 1f, even if the old one was different.
By the time, we propose this PR, you already had change this implementation by muting the master bus.
We still propose individual muting because, well, it is already coded.  But if it troubles you because it seems too complicated or just outside of the initial FMOD API, we can decide to just remove it. It was just a side feature we though was easy to implement with our more general reworking.

Reply:
Mmm you need to introduce new variables, make changes to mute and unmute and add additional update logic to the update method. 

The other changes are all good.

Reply:
Thinking about this from a sound designer's perspective, it's highly unlikely for them to issue a manual setVolume call instead of handing all this in the authoring tool through parameters. It would be nice if the integration abstracts this out but handling it in script for a single event isn't all that bad (although, ideally, you should always stick to parameters).

Reply:
Speaking of sticking to parameters, I've just realized that if the volume is changed through parameters instead of manual setVolume calls, there is no way to "check" for these changes without querying the volume for every event instance per update, which can be very costly.

Reply:
> Our initial reasonning with the "old volume caching" was your first implementation of UnMuteAll() where you was setting the volume to 1f, even if the old one was different.

Yep this was certainly not the right way to do things. 

> We still propose individual muting because, well, it is already coded. But if it troubles you because it seems too complicated or just outside of the initial FMOD API, we can decide to just remove it. 

Given the circumstances i think it would be best if we make a judgement call here and handle it on the master bus without manually trying to do it ourselves. Technically you won't be "losing functionality" as it's outside of the FMOD spec.

Reply:
Well of course, if the volume is handled from the event parameters you should not manually set the volume from the engine. I am not a sound designer so I lack this perspective and might miss lot of obvious things. 
But even if the mute doesn't exist in the API, the setVolume does so it means you can from a game designer's perspective choose the volume you want. It all depend where this responsability lies in your team.  But the topic is not about setVolume, but setMute. 
And yeah you're right, if the volume change with parameters so the oldVolume and IsMuted variables are useless. But in the same condition where you can manually set the volume, you can as well decide to manually mute the event. 
Considering this, I understand why you said that it's just " a nice to have feature". So we will remove individual muting and use the masterBus instead.

Reply:
Regardless of volume/mute.
I still think we need to keep track of OneShot event if we wish to use a setEventPaused.
We might juste use the id returned by the creation function but then we get back to a previous question.
Do we need the verify the validity of an ID inside the integration or is this the responsability of the GdScript side ?

Reply:
You cannot call setPaused on a one shot event. Once you let one loose, that's it. You cannot call functions on it. This is by design. Pausing has to be handled on the bus (eg. an in-game pause). 

If you need to call functions (eg. pausing, setting parameters) you need to create the instance yourself and these instances are not managed by the integration.

Reply:
If the integration doesn't check and tries to use an incorrect ID, the application might just crash (the exact behavior of accessing protected memory depends on the OS/hardware).
If the integration checks, we can abord the function and print a error message, which can be useful for debugging the GdScript side.

Reply:
Well forget everything I said. I totally forgot that OneShot Event doesn't return an ID.

Reply:
Well I delete the possibility to mute a single event.

Reply:
To sum up:
-We remove IsMuted, OldVolume, IsOneShot from the EventInfo.  Only the gameObject and signal dictionnaries remain.
-We remove MuteOneEvent and use the masterBus to pause AllEvent at once.
-We no longer store id of OneShot Event and immediatly call release() when started.
-We simplify the update function. It only need update attached event position, clean up if the gameobject has been deleted, run callback and finally update Fmod.